polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java

```diff
@@ -151,87 +173,195 @@ public void initialize(Bootstrap<PolarisApplicationConfig> bootstrap) {
-      otAware.setOpenTelemetry(openTelemetry);
+  /**
+   * Value instantiator that uses the ServiceLocator to create instances of the various service
+   * types
+   */
+  private static class ServiceLocatorValueInstantiator extends StdValueInstantiator {
+    private final ServiceLocator serviceLocator;
+
+    public ServiceLocatorValueInstantiator(
+        DeserializationConfig config, JavaType valueType, ServiceLocator serviceLocator) {
+      super(config, valueType);
+      this.serviceLocator = serviceLocator;
     }
-    PolarisConfigurationStore configurationStore = configuration.getConfigurationStore();
-    if (metaStoreManagerFactory instanceof ConfigurationStoreAware) {
-      ((ConfigurationStoreAware) metaStoreManagerFactory).setConfigurationStore(configurationStore);
+
+    @Override
+    public boolean canCreateUsingDefault() {
+      return true;
     }
-    RealmEntityManagerFactory entityManagerFactory =
-        new RealmEntityManagerFactory(metaStoreManagerFactory);
-    CallContextResolver callContextResolver = configuration.getCallContextResolver();
-    callContextResolver.setMetaStoreManagerFactory(metaStoreManagerFactory);
-    if (callContextResolver instanceof ConfigurationStoreAware csa) {
-      csa.setConfigurationStore(configurationStore);
+
+    @Override
+    public boolean canInstantiate() {
+      return true;
     }
 
-    RealmContextResolver realmContextResolver = configuration.getRealmContextResolver();
-    realmContextResolver.setMetaStoreManagerFactory(metaStoreManagerFactory);
+    @Override
+    public Object createUsingDefault(DeserializationContext ctxt) throws IOException {
+      return ServiceLocatorUtilities.findOrCreateService(serviceLocator, getValueClass());
+    }
+  }
+
+  /**
+   * Authenticator uses the java Class name in the YAML configuration. Since the Authenticator
+   * interface is owned by Dropwizard, we need to use a MixIn to tell Jackson what field to use to
+   * discover the implementation from the configuration.
+   */
+  @JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, property = "class")
+  static final class NamedAuthenticator {}
+
+  @Override
+  public void run(PolarisApplicationConfig configuration, Environment environment) {
+    OpenTelemetry openTelemetry = setupTracing();
+    PolarisMetricRegistry polarisMetricRegistry =
+        configuration.findService(PolarisMetricRegistry.class);
+
+    MetaStoreManagerFactory metaStoreManagerFactory =
+        configuration.findService(MetaStoreManagerFactory.class);
+
+    // Use the PolarisApplicationConfig to register dependencies in the Jersey resource
+    // configuration. This uses a different ServiceLocator from the one in the bootstrap step
+    environment.jersey().register(configuration.binder());
+    environment
+        .jersey()
+        .register(
+            new AbstractBinder() {
+              @Override
+              protected void configure() {
+                bind(RealmScopeContext.class)
```

Comment:
Looks like there's a lot of manual coding needed for hk2, which feels like it makes things complicated, when we eventually have a plugin/extension mechanism.

Reply:
I think this should be considered the smallest change necessary to get CDI working. There are improvements that can be made, e.g., classpath scanning for detecting Resource and Service impls, but the primary goal right now is to have runtime framework dependent on `@Inject` for wiring. Once that's done, lots of other changes can happen in parallel - e.g., improvements to the authorizer, authentication changes, as well as improvements to the CDI framework. I think this accomplishes the smallest change necessary to allow those other changes to progress.